src/api/mod.rs

```diff
@@ -403,6 +451,119 @@ impl<'a> ValueListBuilder<'a> {
     }
 }
 
+//TODO: move to its own module metadata, with related types
+fn to_meta_data(meta_hm: &HashMap<String, MetaValue>) -> Result<*mut meta_data_t, SubmitError> {
+    let meta;
+    unsafe {
+        meta = meta_data_create();
+    }
+    for (key, value) in meta_hm.iter() {
+        let c_key = CString::new(key.as_str()).map_err(|e| {
+            SubmitError::Field(
+                "meta key",
+                ArrayError::NullPresent(e.nul_position(), key.to_string()),
+            )
+        })?;
+        match value {
+            MetaValue::String(str) => {
+                let c_value = CString::new(str.as_str()).map_err(|e| {
+                    SubmitError::Field(
+                        "meta value",
+                        ArrayError::NullPresent(e.nul_position(), str.to_string()),
+                    )
+                })?;
+                unsafe {
+                    meta_data_add_string(meta, c_key.as_ptr(), c_value.as_ptr());
+                }
+            }
+            MetaValue::SignedInt(i) => unsafe {
+                meta_data_add_signed_int(meta, c_key.as_ptr(), *i);
+            },
+            MetaValue::UnsignedInt(u) => unsafe {
+                meta_data_add_unsigned_int(meta, c_key.as_ptr(), *u);
+            },
+            MetaValue::Double(d) => unsafe {
+                meta_data_add_double(meta, c_key.as_ptr(), *d);
+            },
+            MetaValue::Boolean(b) => unsafe {
+                meta_data_add_boolean(meta, c_key.as_ptr(), *b);
+            },
+        }
+    }
+    Ok(meta)
+}
+
+fn from_meta_data(
+    p: &str,
+    meta: *mut meta_data_t,
+) -> Result<Option<HashMap<String, MetaValue>>, ReceiveError> {
+    if meta.is_null() {
+        return Ok(None);
+    }
+
+    let mut toc: *mut *mut c_char = ptr::null_mut();
+    let count = unsafe { meta_data_toc(meta, &mut toc as *mut *mut *mut c_char) };
+    if count == 0 {
```

Comment:
Since collectd uses less than or zero to signify an error (ref: https://github.com/collectd/collectd/blob/1da3305c10c8ff9a63081284cf3d4bb0f6daffd8/src/utils/format_json/format_json.c#L292) we should do the same thing. That way count can be `usize` and then you can use `toc.add` instead of `offset` which should be more intuitive. 

Even farther: I think you should be able to use `std::slice::from_raw_parts` instead of manually lopping through.

Reply:
i refactored this block to check for an error code, and use slice::from_raw_parts

Reply:
Nice, I think it looks better 👍